engine/src/flutter/lib/web_ui/lib/src/engine/semantics/semantics.dart

```diff
@@ -3057,3 +3107,30 @@ enum EnabledState {
   /// The node is disabled.
   disabled,
 }
+
+extension SemanticsValidationResultExtension on ui.SemanticsValidationResult {
+  /// Returns the ARIA value for this enum value.
+  ///
+  /// The ARIA value can be applied to the `aria-invalid` attribute.
+  String toAriaValue() {
+    switch (this) {
+      case ui.SemanticsValidationResult.none:
+        assert(false, 'ui.SemanticsValidationResult.none should not be converted to ARIA value.');
+        return '';
+      case ui.SemanticsValidationResult.valid:
+        // 'false' may seem counter-intuitive for a "valid" result, but it's
+        // because the ARIA attribute is `aria-invalid`, so its value is
+        // reversed.
+        return 'false';
+      case ui.SemanticsValidationResult.invalid:
+        // 'true' may seem counter-intuitive for an "invalid" result, but it's
+        // because the ARIA attribute is `aria-invalid`, so its value is
+        // reversed.
+        return 'true';
+      case ui.SemanticsValidationResult.grammar:
+        return 'grammar';
```

Comment:
does aria-invalid='grammar' actually work?

Reply:
It works in the sense that the screen reader still reads "invalid entry", but as of today all screen readers I tested on the result is no different from just doing `aria-invalid="true"`.